soc: arm: st_stm32: stm32g4: STM32G4 support

rosterloh opened the pull request:
Add support for STM32G4 series

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh commented:
Requires PR from stm32_hal https://github.com/zephyrproject-rtos/hal_stm32/pull/20

loicpoulain commented:
It would worth to split the PR into several commits, e.g. one for the SoC support, one for the new board(s) support...

rosterloh committed:
soc: arm: st_stm32: Add STM32G4 SoC series

Add soc and dts files to support for most of the common peripherals
in the STM32G4 series. Add specific support for the STM32G431RB.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh commented:
Split the commit as requested by @loicpoulain and added copyright dues zephyrbot warnings

erwango reviewed (commented):
@rosterloh , thanks for this change with this wide range of updated drivers.
Did you had time to test everything?
As mentioned by @loicpoulain, it eases the review when change is split in several small commits.
Here is an example of split PR that was done for L1 port: https://github.com/zephyrproject-rtos/zephyr/pull/12354
I fully understand it is more initial work for author, but since review work is getting easier, upstream goes faster overall.

rosterloh commented:
@erwango I've done the following tests:

erwango commented:
@rosterloh 

> @erwango I've done the following tests:

Ok, then please focus your commits on drivers actually validated: clock, gpio, interrupt_controller.
And remove reference (in doc, dts, driver modifications, ...) for whatever was not validated yet: SPI, I2C, ADC, PWM, UART, ....
These will be added later on, either by you when you have time or by other community members that will take advantage of your initial basic port.

rosterloh commented:
@erwango I've validated everything I've checked in. Just pushed a few fixes in and additional commit. clock, gpio, i2c, interrupt_controller, pwm, led, pinmux, serial, spi and usb all working.

erwango reviewed (changes requested):
@rosterloh, thanks for the work so far, looks promising.
To get merged there is some additional work that will be requested:
- get automatic checks passing (shippable + checkpatch (you can let the ">80 char" warning for now))
- split the PR into subsystem commits (as mentioned already L1 port (#12354) is a good example of what is expected). This is requested to ease the review work and has been followed for all other series porting PRs. 
- Squash the "fix commit" into the previous commits

Thanks

rosterloh commented:
@erwango I can't fix the shippable errors as they require my PR from the hal_stm32 repo to be merged. I've made the other small changes from checkpatch. Do you want me to split the commits further than I already have?

erwango commented:
> @erwango I can't fix the shippable errors as they require my PR from the hal_stm32 repo to be merged. 

Ok, you actually need to push an updated version of west.yml manifest targeting https://github.com/zephyrproject-rtos/hal_stm32/pull/20 (cf: https://docs.zephyrproject.org/latest/contribute/index.html?highlight=modules#contributions-to-external-modules).
This will allow CI to build using the HAL. Then, once we're satisfied with current PR, the PR on hal_stm32 will be merged and we'll get the hal_stm32 SHA1 to update west.yml. And finally current PR with this last west.yml version.

> I've made the other small changes from checkpatch. Do you want me to split the commits further than I already have?

Indeed, please refer to https://docs.zephyrproject.org/latest/contribute/index.html#other-commit-expectations.

rosterloh committed:
ext: hal: st: stm32cube: Add HAL for the STM32G4x series

This commit adds the STM32CUBE HAL for STM32G4x series MCUs.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh committed:
drivers: clock_control: Add STM32G4X clock support

Add clock support for STM32G4X SoC series.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh committed:
drivers: pinmux: Add STM32G4X pinmux support

Add pinmux support for STM32G4X SoC series.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh committed:
drivers: gpio: Add STM32G4X gpio support

Add GPIO driver support for STM32G4X SoC series.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh committed:
drivers: interrupt_controller: Add STM32G4X exti support

Add EXTI driver support for STM32G4X SoC series.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh committed:
drivers: serial: Add STM32G4X serial support

Add UART driver support for STM32G4X SoC series.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh committed:
drivers: flash: Add STM32G4X flash support

Add flash driver support for STM32G4X SoC series.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh committed:
drivers: i2c: Add STM32G4X I2C support

Add I2C driver support for STM32G4X SoC series.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh committed:
drivers: spi: Add STM32G4X SPI support

Add SPI driver support for STM32G4X SoC series.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh committed:
drivers: adc: Add STM32G4X ADC support

Add ADC driver support for STM32G4X SoC series.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh committed:
drivers: counter: Add STM32G4X counter support

Add counter driver support for STM32G4X SoC series.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh committed:
drivers: entropy: Add STM32G4X RNG support

Add RNG driver support for STM32G4X SoC series.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh committed:
drivers: usb: Add STM32G4X USB support

Add USB driver support for STM32G4X SoC series.

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

rosterloh committed:
boards: arm: Add support for Nucleo G431RB board

Nucleo-G431RB board support and description

Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

erwango commented:
@rosterloh, thanks for the split into multiple commits.
Usually, for review efficiency and consistency, I prefer these are split by topic rather by pure directory point of view (eg: usart dts, Kconfig, _fixup, ... into a uart commit, and same for I2C, ...) (as you could see in L1 port (#12354) for instance).
Though, considering the effort made to bring all the drivers, I'll start reviewing in current state.
Meanwhile, if you have time to do this, it would be welcome.

rosterloh commented:
@erwango Thank you for being understanding. I am very busy with my current project at work at the moment but I thought it would be good to push back my changes to help the community. I will try my best to find the time.

erwango committed:
drivers: watchdog: Add STM32G4x Watchdog support

Add watchdog driver support for STM32G4X SoC series.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

erwango committed:
drivers: pwm: Add STM32G4X PWM Support

Add PWM driver support for STM32G4X SoC series.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Richard Osterloh <richard.osterloh@gmail.com>

erwango commented:
@rosterloh , pushed most of the re-split work on you repo, this is a quick work that should be reviewed but I hope you can take it and push it back

rosterloh commented:
Thank you very much @erwango. I really appreciate your help.

erwango commented:
@rosterloh, seems now we have way too much commits (as shown in github, even if not visible on git tree), and I won't be able to help on that front.
I think the easiest to get back on our feet is for you to take back the branch I pushed to your project and push this one directly (and then skip the merge in your branch).
Sorry for additional steps, but this is the first time I tried to work this way.
Thanks

erwango commented on boards/arm/nucleo_g431rb/doc/index.rst:
```diff
@@ -0,0 +1,236 @@
+
+Nucleo G431RB board includes an ST-LINK/V3E embedded debug tool
+interface.  This interface is supported by the openocd version
+included in the Zephyr SDK since v0.10.4.
```
Is it ?
Current Zephyr SDK is 0.10.3. And for now, I don't see any g4 support in openocd upstream (unfortunately)

rosterloh replied:
I submitted this but I don't think it's going to go anywhere https://github.com/zephyrproject-rtos/openocd/pull/11

erwango commented on boards/arm/nucleo_g431rb/doc/index.rst:
```diff
@@ -0,0 +1,236 @@
+Now build and flash an application. Here is an example for
+:ref:`hello_world`.
+
+.. zephyr-app-commands::
```
If upstream openocd cannot be used, please provide alternate flashing method.
I suggest pyocd, which is more flexible for target additions (cf nucleo_wb55rg notice)

rosterloh replied:
Changed to pyocd

erwango commented on boards/arm/nucleo_g431rb/nucleo_g431rb.yaml:
```diff
@@ -0,0 +1,20 @@
+flash: 128
+supported:
+  - arduino_i2c
+  - nvs
```
You can add arduino_spi and arduino_gpio

rosterloh replied:
Done

erwango commented on boards/arm/nucleo_g431rb/nucleo_g431rb_defconfig:
```diff
@@ -0,0 +1,62 @@
+CONFIG_ARM_MPU=y
+
+# Enable pwm
+CONFIG_PWM=y
```
Drivers should not be enabled by default (except, clock, gpio and serial), activation is up to applications.

rosterloh replied:
Removed.

erwango commented on boards/arm/nucleo_g431rb/support/openocd.cfg:
```diff
@@ -0,0 +1,7 @@
+source [find interface/stlink.cfg]
```
Remove this file if not supported

rosterloh replied:
Removed.

erwango commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -176,6 +176,13 @@
 			label = "LPUART_1";
```
Move this commit before board as you configure it in board commit

rosterloh replied:
Could you clarify when you mean by this comment?

erwango replied:
Board configuration will use code provided in drivers commits.
So board commit should be the last one in this series.

rosterloh replied:
It is the last commit on my branch

erwango replied:
Ok, it doesn't appear in that order, this is a github problem then.

erwango commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -261,6 +261,118 @@
+				#pwm-cells = <2>;
+			};
+		};
+
```
timers4 missing ?

rosterloh replied:
Added

erwango commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -261,6 +261,118 @@
+				#pwm-cells = <2>;
+			};
+		};
+
```
TIM17 missing?

rosterloh replied:
Added

erwango commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -261,6 +261,118 @@
+				#pwm-cells = <2>;
+			};
+		};
+
```
TIM8 missing?

rosterloh replied:
Added

erwango commented on dts/arm/st/g4/stm32g4-pinctrl.dtsi:
```diff
@@ -0,0 +1,98 @@
+/*
```
Remove this commit

rosterloh replied:
Done

erwango reviewed (changes requested):
Overall is good, some comments.

dbkinder reviewed (changes requested):
docs look OK, except for the reference to SDK 0.10.4 (latest version is 0.10.3)

rosterloh commented on drivers/clock_control/clock_stm32_ll_common.c:
```diff
@@ -332,7 +340,9 @@ static int stm32_clock_control_init(struct device *dev)
 
 	/* Disable other clocks */
 	LL_RCC_HSE_Disable();
+#ifndef CONFIG_SOC_SERIES_STM32G4X
```
Good idea. Done

rosterloh commented on drivers/pinmux/stm32/pinmux_stm32g4x.h:
```diff
@@ -0,0 +1,407 @@
+#define STM32G4X_PINMUX_FUNC_PA15_I2C1_SCL                                     \
+	(STM32_PINMUX_ALT_FUNC_4 | STM32_OPENDRAIN_PULLUP)
+#define STM32G4X_PINMUX_FUNC_PA15_SPI1_NSS                                     \
+	(STM32_PINMUX_ALT_FUNC_5 | STM32_PUSHPULL_NOPULL)
```
Done

rosterloh commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -102,6 +102,61 @@
+			label = "UART_4";
+		};
+
+		uart5: serial@40005000 {
```
Removed

rosterloh commented on soc/arm/st_stm32/stm32g4/dts_fixup.h:
```diff
@@ -71,4 +71,58 @@
+#define DT_UART_STM32_UART_4_CLOCK_BUS		DT_ST_STM32_UART_40004C00_CLOCK_BUS
+#define DT_UART_STM32_UART_4_HW_FLOW_CONTROL	DT_ST_STM32_UART_40004C00_HW_FLOW_CONTROL
+
+#define DT_UART_STM32_UART_5_BASE_ADDRESS	DT_ST_STM32_UART_40005000_BASE_ADDRESS
```
Done

rosterloh commented on dts/arm/st/g4/stm32g4.dtsi:
```diff
@@ -175,6 +176,58 @@
+			label= "I2C_3";
+		};
+
+		i2c4: i2c@40008400 {
```
Removed

rosterloh commented on soc/arm/st_stm32/stm32g4/dts_fixup.h:
```diff
@@ -189,4 +189,70 @@
 #define DT_SPI_3_CLOCK_BITS			DT_ST_STM32_SPI_FIFO_40003C00_CLOCK_BITS
 #define DT_SPI_3_CLOCK_BUS			DT_ST_STM32_SPI_FIFO_40003C00_CLOCK_BUS
 
+#define DT_PWM_STM32_1_DEV_NAME             	DT_ST_STM32_PWM_40012C00_PWM_LABEL
```
Done

rosterloh commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -182,6 +182,14 @@ static int entropy_stm32_rng_init(struct device *dev)
+		/* Wait for HSI48 to become ready */
+	}
+
+	LL_RCC_SetRNGClockSource(LL_RCC_USB_CLKSOURCE_HSI48);
```
Good catch. Changed.

rosterloh commented on boards/arm/nucleo_g431rb/board.cmake:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+board_runner_args(jlink "--device=STM32G431RB" "--speed=4000")
```
It's in the list [here](https://www.segger.com/downloads/supported-devices.php?model=flasher-portable)

rosterloh commented on boards/arm/nucleo_g431rb/board.cmake:
```diff
@@ -0,0 +1,6 @@
+
+board_runner_args(jlink "--device=STM32G431RB" "--speed=4000")
+
+include(${ZEPHYR_BASE}/boards/common/openocd.board.cmake)
```
Ok

rosterloh commented on soc/arm/st_stm32/stm32g4/soc.h:
```diff
@@ -61,6 +61,12 @@
 #include <stm32g4xx_ll_adc.h>
 #endif
 
+#if defined(CONFIG_RTC_STM32) || defined(CONFIG_COUNTER_RTC_STM32)
```
Removed

erwango commented:
@rosterloh, please rebase your branch as there was an issue with a Kconfig symbol which screwed your CI and is fixed now.
Also, about identity checks, please add our both sign-off to the guilty commit .

rosterloh commented:
Seems like my flash partition for settings is too small now. Is there a minimum?

rosterloh commented:
@erwango on closer inspection ci failing because the application it's building is too big for the FLASh on the device. Is there a mechanism for blacklisting certain tests on certain boards?

erwango commented:
@rosterloh 

Theoretically, CI uses board ram/flash size decalred in board yaml to select which applications could be run on which targets. I wonder why this is failing here.
According to test filters, minimum flash required is 64 and you get 128. There are also other filters that might prevent build on other targets.
Need to find other boards on which it is tested and see flash size. If others targets have similar binary size, issue lies in the test requirements.

erwango commented:
Building disco_l475_iot1, I get
```
[145/156] Linking C executable zephyr/zephyr_prebuilt.elf
Memory region         Used Size  Region Size  %age Used
           FLASH:       67356 B         1 MB      6.42%
            SRAM:       13920 B        96 KB     14.16%
        IDT_LIST:         152 B         2 KB      7.42%

```

While your board gets :
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      132472 B       128 KB    101.07%
            SRAM:       14080 B        32 KB     42.97%
        IDT_LIST:         120 B         2 KB      5.86
```
I wonder how we end up with double flash size on this target

rosterloh commented:
Hmmm, that's really odd. I'm making mods for the gpio api which changed yesterday so maybe the next build will be different.

erwango commented:
@ioannisg , can you help us here?

Seems `tests/arch/arm/arm_thread_swap` is run on board introduced in this PR with no_optimization flavor, which results in  a bit more than 128K flash. 
Test is filtering this no_optim config on > 64 min flash, this sounds optimistic.
Would that be ok to filter on > 128 ?

erwango commented:
> I'm making mods for the gpio api which changed yesterday so maybe the next build will be different.

I'd prefer you don't mix both topics in this PR.
PR is in good shape to be merged soon and there is already a work planned for GPIO API. Your contribution would be of course welcome on this, but please keep it for another PR.

ioannisg commented:
> 
> 
> @ioannisg , can you help us here?
> 
> Seems `tests/arch/arm/arm_thread_swap` is run on board introduced in this PR with no_optimization flavor, which results in a bit more than 128K flash.
> Test is filtering this no_optim config on > 64 min flash, this sounds optimistic.
> Would that be ok to filter on > 128 ?

@erwango sure, I think this is OK.

erwango commented:
@rosterloh, just pushed #19302, when/if accepted please rebase on it to fix shippable issue

dbkinder reviewed (approved):
doc changes LGTM, thanks.

ioannisg commented on soc/arm/st_stm32/stm32g4/soc.h:
```diff
@@ -0,0 +1,84 @@
+ * Therefore, it is essential to include kernel_includes.h after including
+ * core SOC-specific headers.
+ */
+#include <kernel_includes.h>
```
Please, for new boards, follow the changed we have been introducing in patches such as #19357 , i.e. do not included kernel header files in soc.h and, instead, include auto-generated DTS headers

rosterloh replied:
@ioannisg Makes sense. Changed.

ioannisg reviewed (changes requested):
A comment about inclusions in soc.h, as we have been doing for other vendors (and, recently, for stm32, as in #19357)

ioannisg commented on west.yml:
```diff
@@ -59,7 +59,7 @@ manifest:
       path: modules/hal/st
     - name: hal_stm32
-      revision: 272281a1990ec7097f1844778955ed60fe28662a
+      revision: pull/20/head
```
This can't be merged as is, of course; _that_ PR needs to go in first.

ioannisg reviewed (approved):
Looks OK to me, for ./soc/... changes.

erwango commented:
@rosterloh, shippable should work now. You can rebase to get it passed

rosterloh commented:
@erwango Woo Hoo! All passed.

erwango commented:
Approved, but last point will be to update `west.yml` file once https://github.com/zephyrproject-rtos/hal_stm32/pull/20 is merged. 
Tagged `DNM` until this is done.

galak commented:
> Approved, but last point will be to update `west.yml` file once [zephyrproject-rtos/hal_stm32#20](https://github.com/zephyrproject-rtos/hal_stm32/pull/20) is merged.
> Tagged `DNM` until this is done.

Merged the HAL repo change.

rosterloh commented:
@erwango @galak made the west.yaml change
